core/src/services/upyun/core.rs

```diff
@@ -0,0 +1,557 @@
+            self.bucket,
+            percent_encode_path(&path)
+        );
+
+        let req = Request::get(url);
+
+        let mut req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.sign(&mut req).await?;
+
+        self.send(req).await
+    }
+
+    pub async fn info(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let path = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://v0.api.upyun.com/{}/{}",
+            self.bucket,
+            percent_encode_path(&path)
+        );
+
+        let req = Request::head(url);
+
+        let mut req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.sign(&mut req).await?;
+
+        self.send(req).await
+    }
+
+    pub async fn upload(
+        &self,
+        path: &str,
+        size: Option<u64>,
+        args: &OpWrite,
+        body: AsyncBody,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://v0.api.upyun.com/{}/{}",
+            self.bucket,
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::put(&url);
+
+        if let Some(size) = size {
+            req = req.header(header::CONTENT_LENGTH, size.to_string())
+        }
+
+        if let Some(mime) = args.content_type() {
+            req = req.header(header::CONTENT_TYPE, mime)
+        }
+
+        if let Some(pos) = args.content_disposition() {
+            req = req.header(X_UPYUN_CONTENT_DISPOSITION, pos)
+        }
+
+        if let Some(cache_control) = args.cache_control() {
+            req = req.header(X_UPYUN_CACHE_CONTROL, cache_control)
+        }
+
+        // Set body
+        let mut req = req.body(body).map_err(new_request_build_error)?;
+
+        self.sign(&mut req).await?;
+
+        Ok(req)
+    }
+
+    pub async fn delete(&self, path: &str) -> Result<Response<IncomingAsyncBody>> {
+        let path = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://v0.api.upyun.com/{}/{}",
+            self.bucket,
+            percent_encode_path(&path)
+        );
+
+        let req = Request::delete(url);
+
+        let mut req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.sign(&mut req).await?;
+
+        self.send(req).await
+    }
+
+    pub async fn copy(&self, from: &str, to: &str) -> Result<Response<IncomingAsyncBody>> {
+        let from = format!("/{}/{}", self.bucket, build_abs_path(&self.root, from));
+        let to = build_abs_path(&self.root, to);
+
+        let url = format!(
+            "https://v0.api.upyun.com/{}/{}",
+            self.bucket,
+            percent_encode_path(&to)
+        );
+
+        let mut req = Request::put(url);
+
+        req = req.header(X_UPYUN_MOVE_SOURCE, from);
```

Comment:
Thanks for the doc address you provided.
But for copy and move, they use different header: X-Upyun-Copy-Source and X-Upyun-Move-Source.
Please check this.

Reply:
thanks, fixed